File: seqspec/seqspec_init.py
"""Init module for seqspec CLI.

This module provides functionality to generate new seqspec files from a newick tree format.
"""

from argparse import ArgumentParser, Namespace, RawTextHelpFormatter
from pathlib import Path
from typing import List

from seqspec.Assay import Assay
from seqspec.Region import Region
from seqspec.utils import write_pydantic_to_file_or_stdout

# example
# seqspec init -n myassay -m 1 -o spec.yaml "(((barcode:16,umi:12)r1.fastq.gz,(cdna:150)r2.fastq.gz)rna)" # seqspec init -n myassay -m 1 -o spec.yaml -r "rna,R1.fastq.gz,truseq_r1,16,pos:rna,R2.fastq.gz,truseq_r2,100,neg" " ((truseq_r1:10,barcode:16,umi:12,cdna:150)rna)"
# seqspec init -n myassay -m 2 -o spec.yaml "(((barcode:16,umi:12)r1.fastq.gz,(cdna:150)r2.fastq.gz)rna,((barcode:16)r1.fastq.gz,(gdna:150)r2.fastq.gz,(gdna:150)r3.fastq.gz)atac)"


def setup_init_args(parser) -> ArgumentParser:
    """Create and configure the init command subparser (simple draft creator)."""
    subparser = parser.add_parser(
        "init",
        description="""
Generate a new *empty* seqspec draft (meta-Regions only).

Example:
    seqspec init -n myassay -m rna,atac -o spec.yaml
""",
        help="Generate a new empty seqspec file",
        formatter_class=RawTextHelpFormatter,
    )
    req = subparser.add_argument_group("required arguments")
    req.add_argument("-n", "--name", required=True, help="Assay name")
    req.add_argument(
        "-m",
        "--modalities",
        required=True,
        help="Comma-separated list of modalities (e.g. rna,atac)",
    )
    subparser.add_argument("--doi", default="", help="DOI of the assay")
    subparser.add_argument("--description", default="", help="Short description")
    subparser.add_argument("--date", default="", help="Date (YYYY-MM-DD)")
    subparser.add_argument(
        "-o", "--output", type=Path, metavar="OUT", help="Output YAML (default stdout)"
    )
    return subparser


def validate_init_args(
    _: ArgumentParser, args: Namespace
) -> None:  # simple draft needs no extra checks
    if not args.name:
        raise ValueError("Assay name is required")
    if not args.modalities:
        raise ValueError("Modalities must be provided")


def run_init(parser: ArgumentParser, args: Namespace) -> None:
    """Run the simplified init command."""
    validate_init_args(parser, args)

    modalities = [m.strip() for m in args.modalities.split(",") if m.strip()]
    spec = seqspec_init(args.name, args.doi, args.date, args.description, modalities)
    # Build empty assay with meta Regions only

    write_pydantic_to_file_or_stdout(spec, args.output)


def seqspec_init(
    name: str, doi: str, date: str, description: str, modalities: List[str]
) -> Assay:
    meta_regions: List[Region] = [
        Region(
            region_id=mod, region_type="meta", name=mod, sequence_type="", regions=[]
        )
        for mod in modalities
    ]

    spec = Assay(
        assay_id="",
        name=name,
        doi=doi,
        date=date,
        description=description,
        modalities=modalities,
        lib_struct="",
        sequence_protocol=None,
        sequence_kit=None,
        library_protocol=None,
        library_kit=None,
        sequence_spec=[],
        library_spec=meta_regions,
    )
    spec.update_spec()
    return spec

File: src/seqspec_init.rs
use crate::models::assay::Assay;
use crate::models::region::Region;
use clap::Args;
use std::fs;
use std::io::Write;
use std::path::PathBuf;

#[derive(Debug, Args)]
pub struct InitArgs {
    #[clap(short, long, help = "Assay name", required = true)]
    name: String,

    #[clap(
        short,
        long,
        help = "Comma-separated list of modalities (e.g. rna,atac)",
        required = true
    )]
    modalities: String,

    #[clap(long, help = "DOI of the assay", default_value = "")]
    doi: String,

    #[clap(long, help = "Short description", default_value = "")]
    description: String,

    #[clap(long, help = "Date (YYYY-MM-DD)", default_value = "")]
    date: String,

    #[clap(short, long, help = "Output YAML (default stdout)", value_name = "OUT")]
    output: Option<PathBuf>,
}

pub fn run_init(args: &InitArgs) {
    validate_init_args(args);

    let modalities: Vec<String> = args
        .modalities
        .split(',')
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect();

    let mut spec = seqspec_init(
        &args.name,
        &args.doi,
        &args.date,
        &args.description,
        modalities,
    );

    spec.update_spec();

    let yaml = spec.to_bytes().expect("Failed to serialize assay to YAML");
    if let Some(out) = &args.output {
        let mut f = fs::File::create(out).unwrap();
        f.write_all(&yaml).unwrap();
    } else {
        println!("{}", String::from_utf8_lossy(&yaml));
    }
}

fn validate_init_args(args: &InitArgs) {
    if args.name.is_empty() {
        eprintln!("Assay name is required");
        std::process::exit(1);
    }
    if args.modalities.is_empty() {
        eprintln!("Modalities must be provided");
        std::process::exit(1);
    }
}

pub fn seqspec_init(
    name: &str,
    doi: &str,
    date: &str,
    description: &str,
    modalities: Vec<String>,
) -> Assay {
    let meta_regions: Vec<Region> = modalities
        .iter()
        .map(|modality| {
            Region::new(
                modality.clone(),   // region_id
                "meta".to_string(), // region_type
                modality.clone(),   // name
                "".to_string(),     // sequence_type
                "".to_string(),     // sequence
                0,                  // min_len
                0,                  // max_len
                None,               // onlist
                Vec::new(),         // regions
            )
        })
        .collect();

    Assay::new(
        "".to_string(), // assay_id
        name.to_string(),
        doi.to_string(),
        date.to_string(),
        description.to_string(),
        modalities,
        "".to_string(), // lib_struct
        Vec::new(),     // sequence_spec
        meta_regions,   // library_spec
        None,           // sequence_protocol
        None,           // sequence_kit
        None,           // library_protocol
        None,           // library_kit
        None,           // seqspec_version
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_init_creates_assay() {
        let spec = seqspec_init(
            "TestAssay",
            "10.1234/test",
            "2024-01-01",
            "A test assay",
            vec!["rna".into(), "atac".into()],
        );
        assert_eq!(spec.name, "TestAssay");
        assert_eq!(spec.doi, "10.1234/test");
        assert_eq!(spec.modalities, vec!["rna", "atac"]);
        assert_eq!(spec.library_spec.len(), 2);
        assert!(spec.sequence_spec.is_empty());
    }

    #[test]
    fn test_init_library_spec_regions() {
        let spec = seqspec_init("Test", "", "", "", vec!["rna".into()]);
        let lib = spec.get_libspec("rna").unwrap();
        assert_eq!(lib.region_id, "rna");
        assert_eq!(lib.region_type, "meta");
    }

    #[test]
    fn test_init_single_modality() {
        let spec = seqspec_init("Test", "", "", "", vec!["protein".into()]);
        assert_eq!(spec.modalities.len(), 1);
        assert_eq!(spec.library_spec.len(), 1);
        assert_eq!(spec.library_spec[0].region_id, "protein");
    }
}
